wrangler-core/src/main/java/io/cdap/wrangler/executor/RecipePipelineExecutor.java

```diff
@@ -161,4 +188,55 @@ private List<Directive> getDirectives() throws RecipeException {
     }
     return directives;
   }
+
+  private Schema generateOutputSchema(Schema inputSchema, List<Row> output) throws RecordConvertorException {
+    Map<String, Schema> outputFieldMap = new LinkedHashMap<>();
+    for (Row row : output) {
+      for (Pair<String, Object> rowField : row.getFields()) {
+        String fieldName = rowField.getFirst();
+        Object fieldValue = rowField.getSecond();
+
+        Schema existing = inputSchema.getField(fieldName) != null ? inputSchema.getField(fieldName).getSchema() : null;
+        Schema generated = fieldValue != null && !isValidSchemaForValue(existing, fieldValue) ?
+          generator.getSchema(fieldValue, fieldName) : null;
+
+        if (generated != null) {
+          outputFieldMap.put(fieldName, generated);
```

Comment:
Does it mean you always use the schema from the last non-null row? If so, there is no reason to run generator many times